File: android/src/main/java/us/mediagrid/capacitorjs/plugins/nativeaudio/AudioMetadata.java
package us.mediagrid.capacitorjs.plugins.nativeaudio;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.getcapacitor.plugin.util.HttpRequestHandler;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

public class AudioMetadata {

    private static final String TAG = "AudioMetadata";

    public String albumTitle;
    public String artistName;
    public String songTitle;
    public String artworkSource;
    public String updateUrl;
    public Integer updateInterval = 15;

    private String onMetadataUpdateCallbackId;

    private Handler updateHandler = null;
    private Runnable updateRunner = null;
    private Runnable updateCallback = null;
    private JSObject updateFullResponse;

    private AudioPlayerPlugin pluginOwner;

    AudioMetadata(
        String albumTitle,
        String artistName,
        String songTitle,
        String artworkSource,
        String updateUrl,
        Integer updateInterval
    ) {
        this.albumTitle = albumTitle;
        this.artistName = artistName;
        this.songTitle = songTitle;
        this.artworkSource = artworkSource;
        this.updateUrl = updateUrl;

        if (updateInterval != null) {
            this.updateInterval = updateInterval;
        }
    }

    public void update(AudioMetadata metadata) {
        albumTitle = metadata.albumTitle;
        artistName = metadata.artistName;
        songTitle = metadata.songTitle;
        artworkSource = metadata.artworkSource;
    }

    public AudioMetadata setPluginOwner(AudioPlayerPlugin plugin) {
        pluginOwner = plugin;

        return this;
    }

    public AudioMetadata setUpdateCallBack(Runnable callback) {
        updateCallback = callback;

        return this;
    }

    public void setOnMetadataUpdate(String callbackId) {
        onMetadataUpdateCallbackId = callbackId;
    }

    public void startUpdater() {
        if (!hasUpdateUrl() || updateHandler != null) {
            return;
        }

        Log.i(TAG, "Starting metadata updater...");

        updateHandler = new Handler(Looper.getMainLooper());
        updateRunner = new Runnable() {
            @Override
            public void run() {
                updateMetadataByUrl(() -> updateHandler.postDelayed(this, updateInterval * 1000));
            }
        };

        updateHandler.post(updateRunner);
    }

    public void stopUpdater() {
        if (updateHandler == null) {
            return;
        }

        Log.i(TAG, "Stopping metadata updater...");

        updateHandler.removeCallbacks(updateRunner);
        updateHandler = null;
        updateRunner = null;
    }

    public boolean hasUpdateUrl() {
        return updateUrl != null && updateUrl != "";
    }

    public void updateMetadataByUrl(Runnable requeueCallback) {
        if (pluginOwner.executorService.isShutdown()) {
            return;
        }

        pluginOwner.executorService.submit(() -> {
            try {
                if (!makeUpdateRequest()) {
                    return;
                }

                if (updateCallback != null) {
                    // Updating the MediaController needs to be on the main thread
                    new Handler(Looper.getMainLooper()).post(() -> {
                        updateCallback.run();
                    });
                }

                if (onMetadataUpdateCallbackId != null) {
                    PluginCall call = pluginOwner
                        .getBridge()
                        .getSavedCall(onMetadataUpdateCallbackId);

                    if (call != null) {
                        call.resolve(updateFullResponse);
                    }
                }
            } catch (Exception ex) {
                Log.e(TAG, "There was an error running the metadata update", ex);
            }

            if (requeueCallback != null) {
                requeueCallback.run();
            }
        });
    }

    private boolean makeUpdateRequest() {
        Log.i(TAG, "Getting metadata from URL " + updateUrl);
        HttpURLConnection urlConnection = null;

        try {
            URL url = new URL(updateUrl);
            urlConnection = (HttpURLConnection) url.openConnection();
            urlConnection.setRequestProperty("Accept", "application/json");

            InputStream errorStream = urlConnection.getErrorStream();

            if (errorStream != null) {
                Log.e(
                    TAG,
                    String.format(
                        "The metadata update server returned a status of %s with the message %s",
                        urlConnection.getResponseCode(),
                        HttpRequestHandler.readStreamAsString(errorStream)
                    )
                );
            } else {
                JSObject json = new JSObject(
                    HttpRequestHandler.readStreamAsString(urlConnection.getInputStream())
                );

                Log.i(TAG, json.toString());

                updateFullResponse = json;
                albumTitle = json.getString("album_title");
                artistName = json.getString("artist_name");
                songTitle = json.getString("song_title");
                artworkSource = json.getString("artwork_source");

                return true;
            }
        } catch (Exception ex) {
            Log.e(TAG, "An error occurred trying to get updated metadata", ex);
        } finally {
            if (urlConnection != null) {
                urlConnection.disconnect();
            }
        }

        return false;
    }
}

File: ios/Sources/AudioPlayerPlugin/AudioMetadata.swift
import Capacitor

public class AudioMetadata {
    var albumTitle: String
    var artistName: String
    var songTitle: String
    var artworkSource: String
    var updateUrl: String
    var updateInterval: Int = 15

    private var onMetadataUpdateCallbackId: String = ""

    private var updateHandler: DispatchSourceTimer!
    private var updateCallback: (() -> Void)!
    private var updateFullResponse: [String: Any] = [:]

    private var pluginOwner: AudioPlayerPlugin?

    public init(
        albumTitle: String,
        artistName: String,
        songTitle: String,
        artworkSource: String,
        updateUrl: String,
        updateInterval: Int
    ) {
        self.albumTitle = albumTitle
        self.artistName = artistName
        self.songTitle = songTitle
        self.artworkSource = artworkSource
        self.updateUrl = updateUrl

        if updateInterval != -1 {
            self.updateInterval = updateInterval
        }
    }

    public func update(metadata: AudioMetadata) {
        self.albumTitle = metadata.albumTitle
        self.artistName = metadata.artistName
        self.songTitle = metadata.songTitle
        self.artworkSource = metadata.artworkSource
    }

    public func setPluginOwner(pluginOwner: AudioPlayerPlugin) -> Self {
        self.pluginOwner = pluginOwner

        return self
    }

    public func setUpdateCallback(callback: @escaping () -> Void) -> Self {
        self.updateCallback = callback

        return self
    }

    public func setOnMetadataUpdate(callbackId: String) {
        onMetadataUpdateCallbackId = callbackId
    }

    public func startUpdater() {
        if !hasUpdateUrl() || updateHandler != nil {
            return
        }

        print("Starting metadata updater...")

        updateHandler = DispatchSource.makeTimerSource(
            queue: DispatchQueue.global(qos: .background)
        )
        updateHandler.schedule(
            deadline: .now(),
            repeating: .seconds(updateInterval),
            leeway: .milliseconds(250)
        )
        updateHandler.setEventHandler {
            self.makeUpdateRequest()
        }

        updateHandler.activate()
    }

    public func stopUpdater() {
        if updateHandler == nil {
            return
        }

        print("Stopping metadata updater...")

        updateHandler.cancel()
        updateHandler = nil
    }

    public func hasUpdateUrl() -> Bool {
        return !updateUrl.isEmpty
    }

    public func updateMetadataByUrl() {
        makeUpdateRequest()
    }

    private func makeUpdateRequest() {
        guard
            let updateUrl = URL.init(string: self.updateUrl)
        else {
            print("Update metadata URL is invalid")
            return
        }

        var updateRequest = URLRequest(url: updateUrl)
        updateRequest.addValue("application/json", forHTTPHeaderField: "Accept")

        print("Getting metadata from URL \(self.updateUrl)")

        URLSession.shared.dataTask(with: updateRequest) { (data, response, error) in
            if let error = error {
                print(error)
                return
            }

            guard let httpResponse = response as? HTTPURLResponse else {
                print("The metadata update server response is invalid")
                return
            }

            guard (200...299).contains(httpResponse.statusCode) else {
                print(
                    "The metadata update server returned a non-2xx status code: \(httpResponse.statusCode)"
                )
                return
            }

            do {
                guard
                    let json = (try JSONSerialization.jsonObject(with: data!)) as? [String: Any]
                else {
                    print("The metadata update data could not be parsed as JSON")
                    return
                }

                print(json)

                self.updateFullResponse = json
                self.albumTitle = json["album_title"] as? String ?? ""
                self.artistName = json["artist_name"] as? String ?? ""
                self.songTitle = json["song_title"] as? String ?? ""
                self.artworkSource = json["artwork_source"] as? String ?? ""
            } catch {
                print(
                    "An error occurred trying to get updated metadata: \(error.localizedDescription)"
                )
            }

            if self.onMetadataUpdateCallbackId != "" {
                DispatchQueue.main.async {
                    self.pluginOwner?.bridge?.savedCall(withID: self.onMetadataUpdateCallbackId)?
                        .resolve(self.updateFullResponse)
                }

            }

            if self.updateCallback != nil {
                DispatchQueue.main.async {
                    self.updateCallback()
                }
            }
        }.resume()
    }
}
